consensus/src/lib.rs

```diff
@@ -15,15 +15,54 @@
 
 //! A consensus related logic.
 
+mod error;
 mod pos;
 mod pow;
+mod validator;
+
+use chainstate_types::{BlockIndex, GenBlockIndex, PropertyQueryError};
+use common::{
+    chain::block::{consensus_data::PoWData, ConsensusData},
+    chain::{Block, ChainConfig, RequiredConsensus},
+    primitives::{BlockHeight, Id},
+};
 
 pub use crate::{
     error::ConsensusVerificationError,
     pos::{check_pos_hash, error::ConsensusPoSError, kernel::get_kernel_output},
-    pow::{check_proof_of_work, mine, ConsensusPoWError},
+    pow::{calculate_work_required, check_proof_of_work, mine, ConsensusPoWError},
     validator::validate_consensus,
 };
 
-mod error;
-mod validator;
+#[allow(unreachable_code)]
+pub fn initialize_consensus_data<F, G>(
+    chain_config: &ChainConfig,
+    block: &mut Block,
```

Comment:
I think it would be better not to pass mut Block here at all. At this point it's not fully constructed which cannot be known from this function's point of view.

Rather this function should accept `height` and `prev_block_id` which are actually utilized here. And return `ConsensusData` as a result. I think it's possible but arguments of few functions need to be changed

Reply:
The `initialize_consensus_data()` function will be used along with a new `finalize_consensus_data()` function, which is where mining will be performed.

But I guess `initialize_consensus_data()` could instead be something like `generate_consensus_data()` which returned `ConsensusData` and then `block.update_consensus_data()` could be moved to outside of the function... I just wanted to hide away the details and push them into the `consensus` crate. However I think you're right - returning `ConsensusData` is a better approach.

As for mining in `finalize_consensus_data()`, I'm still thinking this should be hidden away inside the `consensus` crate, but maybe you'll have a better approach once I push a PR soon.

Reply:
Ok, this turned out nice 👍

Reply:
I had a look and I'm not sure about this one. By using BlockHeader, it's showing at the interface level what the meaning of that parameter is. Otherwise, we would need 3 individual fields to be passed in separately, and then we lose that grouping meaning. Don't know, maybe I'm wrong here?